File: frontend/src/api/lineModels.ts
import { apiClient } from './client'

export interface LineModel {
  line_model_id: string
  line_type: string
  model_name: string
  r_ohm_per_km: number
  x_ohm_per_km: number
  c_nf_per_km: number
  source: 'preset' | 'custom'
  enabled: boolean
  remark: string | null
}

export interface LineModelCreatePayload {
  line_type: string
  model_name: string
  r_ohm_per_km: number
  x_ohm_per_km: number
  c_nf_per_km: number
  remark?: string | null
}

export interface LineModelUpdatePayload {
  model_name?: string
  r_ohm_per_km?: number
  x_ohm_per_km?: number
  c_nf_per_km?: number
  enabled?: boolean
  remark?: string | null
}

export async function listLineModels(params?: { enabled_only?: boolean; line_type?: string }): Promise<LineModel[]> {
  const { data } = await apiClient.get<LineModel[]>('/line-models', { params })
  return data
}

export async function createLineModel(payload: LineModelCreatePayload): Promise<LineModel> {
  const { data } = await apiClient.post<LineModel>('/line-models', payload)
  return data
}

export async function updateLineModel(id: string, payload: LineModelUpdatePayload): Promise<LineModel> {
  const { data } = await apiClient.patch<LineModel>(`/line-models/${id}`, payload)
  return data
}

export async function deleteLineModel(id: string): Promise<void> {
  await apiClient.delete(`/line-models/${id}`)
}

File: backend/app/api/line_models.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from app.db.session import get_db
from app.models import LineModel
from app.schemas.line_model import LineModelCreate, LineModelOut, LineModelUpdate
from app.services.line_model_service import (
    LineModelInUseError,
    LineModelNotFoundError,
    create_line_model,
    delete_line_model,
    list_line_models,
    update_line_model,
)

router = APIRouter(prefix="/api/line-models", tags=["line-models"])


@router.get("", response_model=list[LineModelOut])
def list_line_models_endpoint(
    enabled_only: bool = Query(False),
    line_type: str | None = Query(None),
    db: Session = Depends(get_db),
) -> list[LineModel]:
    return list_line_models(db, enabled_only=enabled_only, line_type=line_type)


@router.post("", response_model=LineModelOut)
def create_line_model_endpoint(payload: LineModelCreate, db: Session = Depends(get_db)) -> LineModel:
    return create_line_model(db, payload)


@router.patch("/{line_model_id}", response_model=LineModelOut)
def update_line_model_endpoint(
    line_model_id: str, payload: LineModelUpdate, db: Session = Depends(get_db)
) -> LineModel:
    try:
        return update_line_model(db, line_model_id, payload)
    except LineModelNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc


@router.delete("/{line_model_id}", status_code=204)
def delete_line_model_endpoint(line_model_id: str, db: Session = Depends(get_db)) -> None:
    try:
        delete_line_model(db, line_model_id)
    except LineModelNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except LineModelInUseError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc
